### PR TITLE
Twitter OAuth2

### LINKED ISSUE #18192: Upgrade Twitter OAuth 1.0a to OAuth 2
We've noticed some issues regarding registration of new users that used Twitter social auth. We have a hunch that upgrading to [OAuth 2.0](https://developer.twitter.com/en/docs/authentication/oauth-2-0) will get rid of these inconsistencies, on top of other benefits this upgrade would bring.

This is a sensitive area of the codebase and also a somewhat complex one too IMO. I'm leaving it with the tag `ready for dev` for now in case anyone wants to take this on, but it will otherwise (likely) be prioritized and assigned to a core maintainer soon.

For reference there's [this DEV post](https://dev.to/unasuke/omniauth-twitter2-gem-how-to-authenticate-twitter-account-by-oauth-20-on-your-rails-app-5fm7) that mentions [omniauth-twitter2](https://github.com/unasuke/omniauth-twitter2) gem as a valid replacement. 

The most important things to keep in mind are likely the structure of the `auth_data_dump` (how we extract the profile data from it) and ensuring existing users can continue to use the app seamlessly.

### PR DESCRIPTION
**Pull request by fdocr**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Optimization

## Related Tickets & Documents

- Closes #18192

## QA Instructions, Screenshots, Recordings

1. Boot app locally
2. Enabled `:twitter_oauth2` FeatureFlag
3. Go to developers.twitter.com and create an app that allows OAuth2 login ([more details on what this looks like here](https://github.com/forem/forem/issues/18192#issuecomment-1353670887))
4. Fill in Client ID and Client Secret on the admin dashboard on `/admin/customization/config`
5. You should be able to sign in/up with Twitter and also connect an existing account via `/settings`

### UI accessibility concerns?

None - the UI remains the same

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] This change does not need to be communicated, and this is why not: We are releasing this new Twitter OAuth option behind a Feature Flag to test this on a production environment and once confirmed we are good to remove the Feature Flag we will communicate the change appropriately

## [optional] Are there any post deployment tasks we need to perform?

Test this on Canary/CodeNewbie community

## [optional] What gif best describes this PR or how it makes you feel?

upgrade - lemonade stand

### CODE REVIEW

File: app/views/shared/authentication/_providers_registration_form.html.erb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <div class="registration__actions-providers">
   <% authentication_enabled_providers.each do |provider| %>
+  <% binding.break %>
```
Comment (fdocr) [inline app/views/shared/authentication/_providers_registration_form.html.erb:3]: Still debugging an issue with a previously enabled "twitter" provider stored in `Settings::Authentication.providers`

**Comment by tyvdh** [comment]
Getting errors with the current self hosted Forem and the Twitter OAuth flow. Would this PR resolve those? Current docs seem quite outdated and with all the changes at Twitter recently I’m unsure where the error may be rooted. 

89E5DE14-2048-4A18-9FC9-B54F6D4ABCD2

**Comment by mirie** [comment]
@fdocr How is this draft PR coming? Do you want feedback prior to getting an "official" PR review?

**Comment by fdocr** [comment]
> How is this draft PR coming? Do you want feedback prior to getting an "official" PR review?

Hey @mirie, I think that would be a great idea. I pushed the progress from late last week & today (force pushed since I had to rewrite a good chunk of it) that included a FeatureFlag to toggle between OAuth2 and the old implementation.

@maestromac could you help me deploy this to a Canary instance to test out on a real instance please? In case it doesn't work there perhaps we could talk it through over a 10-15min call (or over Slack too) sometime tomorrow or later this week since I occasionally encounter some issues with the configs 🙂

**Comment by fdocr** [comment]
Hi @tyvdh, thanks for reaching out on that problem you're seeing. That's odd since the current implementation should continue to work as expected (it's what we currently use on all existing Forem instances like https://dev.to).

The docs are indeed outdated, but the overall process remains the same. You should enable the Twitter authentication method within `/admin/customization/config` with the OAuth 1.0a Key and Secret you get from https://developer.twitter.com/

I would recommend (1) double checking your key/secret, (2) your callback URL configured in the Twitter developer portal, and (3) that the twitter account username you're using is indeed 1-30 characters long.

If that doesn't work you can open an issue in the repo and we can help troubleshoot there. This PR is an enhancement to the current working implementation, so to answer your question, I don't believe this is the root cause of your problem from what you originally mentioned. Hope this helps!

**Comment by mirie** [comment]
> @maestromac could you help me deploy this to a Canary instance to test out on a real instance please? In case it doesn't work there perhaps we could talk it through over a 10-15min call (or over Slack too) sometime tomorrow or later this week since I occasionally encounter some issues with the configs 🙂

@maestromac Can you also make sure that Fernando can learn how to do this himself for the future? I have a request to deploy https://github.com/forem/forem/issues/18649 to a Canary instance too for QA and this would be a good opportunity to immediately apply what he learned 😄  -- can we coordinate in #eng both of these deployments (possibly today?)

@fdocr Do you want QA feedback or technical feedback on this PR? Can you tag in a comment for technical feedback?

**Comment by mirie** [comment]
Talked with @fdocr about my question above:

1. @fdocr is going to do some QA on Canary 2
2. If he needs support will work with @maestromac 
3. Once he's satisfied with the state of the work, will move this out of draft and get technical review from team via PR review process

**Review by Ridhwana — APPROVED** [review]

**Review by rt4914 — COMMENTED** [review]
I am unable to test this locally as twitter sends a verification link on gmail but its actually not visible in inbox or spam after multiple attempts.

I only cases that I want to test are the once which are failing in this comment
https://github.com/forem/forem/issues/18192#issuecomment-1370529284


If someone can have a look that would be great.

**Comment by mirie** [comment]
Closing for now. We might pick this up again in the future.

**Pull request closed without merging**